cmd/import-validator/core/service.go

```diff
@@ -128,102 +128,142 @@ func openLedger(ledgerPath string, genesis *bookkeeping.Genesis, genesisBlock *b
 	return ledger, nil
 }
 
-func getModifiedAccounts(l *ledger.Ledger, block *bookkeeping.Block) ([]basics.Address, error) {
+func getModifiedState(l *ledger.Ledger, block *bookkeeping.Block) (map[basics.Address]struct{}, map[basics.Address]map[ledger.Creatable]struct{}, error) {
 	eval, err := l.StartEvaluator(block.BlockHeader, len(block.Payset), 0)
 	if err != nil {
-		return nil, fmt.Errorf("changedAccounts() start evaluator err: %w", err)
+		return nil, nil, fmt.Errorf("getModifiedState() start evaluator err: %w", err)
 	}
 
 	paysetgroups, err := block.DecodePaysetGroups()
 	if err != nil {
-		return nil, fmt.Errorf("changedAccounts() decode payset groups err: %w", err)
+		return nil, nil, fmt.Errorf("getModifiedState() decode payset groups err: %w", err)
 	}
 
 	for _, group := range paysetgroups {
 		err = eval.TransactionGroup(group)
 		if err != nil {
-			return nil, fmt.Errorf("changedAccounts() apply transaction group err: %w", err)
+			return nil, nil,
+				fmt.Errorf("getModifiedState() apply transaction group err: %w", err)
 		}
 	}
 
 	vb, err := eval.GenerateBlock()
 	if err != nil {
-		return nil, fmt.Errorf("changedAccounts() generate block err: %w", err)
+		return nil, nil, fmt.Errorf("getModifiedState() generate block err: %w", err)
 	}
 
-	accountDeltas := vb.Delta().Accts
-	return accountDeltas.ModifiedAccounts(), nil
+	accountDeltas := vb.Delta().NewAccts
+
+	modifiedAccounts := make(map[basics.Address]struct{})
+	for _, address := range accountDeltas.ModifiedAccounts() {
+		modifiedAccounts[address] = struct{}{}
+	}
+
+	modifiedResources := make(map[basics.Address]map[ledger.Creatable]struct{})
+	for _, r := range accountDeltas.GetAllAssetResources() {
+		c, ok := modifiedResources[r.Addr]
+		if !ok {
+			c = make(map[ledger.Creatable]struct{})
+			modifiedResources[r.Addr] = c
+		}
+		creatable := ledger.Creatable{
+			Index: basics.CreatableIndex(r.Aidx),
+			Type:  basics.AssetCreatable,
+		}
+		c[creatable] = struct{}{}
+	}
+	for _, r := range accountDeltas.GetAllAppResources() {
+		c, ok := modifiedResources[r.Addr]
+		if !ok {
+			c = make(map[ledger.Creatable]struct{})
+			modifiedResources[r.Addr] = c
+		}
+		creatable := ledger.Creatable{
+			Index: basics.CreatableIndex(r.Aidx),
+			Type:  basics.AppCreatable,
+		}
+		c[creatable] = struct{}{}
+	}
+
+	return modifiedAccounts, modifiedResources, nil
+}
+
+func normalizeAccountResource(r ledgercore.AccountResource) ledgercore.AccountResource {
+	if (r.AppParams != nil) && (len(r.AppParams.GlobalState) == 0) {
+		// Make a copy of `AppParams` to avoid modifying ledger's storage.
+		appParams := new(basics.AppParams)
+		*appParams = *r.AppParams
+		appParams.GlobalState = nil
+		r.AppParams = appParams
+	}
+	if (r.AppLocalState != nil) && (len(r.AppLocalState.KeyValue) == 0) {
+		// Make a copy of `AppLocalState` to avoid modifying ledger's storage.
+		appLocalState := new(basics.AppLocalState)
+		*appLocalState = *r.AppLocalState
+		appLocalState.KeyValue = nil
```

Comment:
go-algorand code can return both empty and nil maps, but indexer code only returns nil maps. So here if the map is empty, we set it to nil, so that comparison doesn't fail.